portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java

```diff
@@ -36,22 +36,40 @@
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/");
+      return;
+    }
+
     String title = request.getParameter("comment-title");
     String author = request.getParameter("name");
     long currentTime = System.currentTimeMillis();
     String comment = request.getParameter("textfield");
     long rating = Long.parseLong(request.getParameter("rating"));
+    String userId = userService.getCurrentUser().getUserId();
+    Key userKey = getUserKey(userId);
 
-    Entity commentEntity = new Entity("Comment");
+    Entity commentEntity = new Entity("Comment", userKey);
     commentEntity.setProperty("title", title);
     commentEntity.setProperty("author", author);
     commentEntity.setProperty("timestamp", currentTime);
     commentEntity.setProperty("comment", comment);
     commentEntity.setProperty("rating", rating);
+    commentEntity.setProperty("userId", userId);
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
 
     response.sendRedirect("/contact.html");
   }
+
+  private Key getUserKey(String userId) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("User")
```

Comment:
[Optional] I personally find it more readable to create a separate variable to store the filter but either way is fine.